sdks/go/pkg/beam/core/runtime/exec/coder.go

```diff
@@ -1088,25 +1121,22 @@ func EncodeWindowedValueHeader(enc WindowEncoder, ws []typex.Window, t typex.Eve
 	if err := enc.Encode(ws, w); err != nil {
 		return err
 	}
-	_, err := w.Write(paneNoFiring)
+	err := coder.EncodePane(p, w)
 	return err
 }
 
 // DecodeWindowedValueHeader deserializes a windowed value header.
-func DecodeWindowedValueHeader(dec WindowDecoder, r io.Reader) ([]typex.Window, typex.EventTime, error) {
+func DecodeWindowedValueHeader(dec WindowDecoder, r io.Reader) ([]typex.Window, typex.EventTime, typex.PaneInfo, error) {
 	// Encoding: Timestamp, Window, Pane (header) + Element
-
+	pn := typex.PaneInfo{}
```

Comment:
It would be more idiomatic to declare the zero value in place for the returns, which makes it explicit that they're the zero value. 
The reason for this is to limit scope, and make it easier on readers. If possible we keep that as small as possible. With a pre-declaration like this, a reader has to keep an eye out for it, for when it's used, and what might have affected it before it's returned. By writing our intent (return a zero value on error) directly where it's happening, we avoid this.

That said, I understand the repetition can be tiresome. You can instead have defined in this function scope helper function

`onError := func(err error) { return  nil, mtime.ZeroTimestamp, typex.PaneInfo{}, err }`

Which you can then call as `return onError(err)` and it will compile down to the same thing.

Non-pointer zeros being inconvenient is a hot topic of discussion for Go and eventually someone will propose a satisfying solution for it.

Reply:
This is great!